File: Blatt_1/3-csv.py
import argparse
import sys
import numpy as np

def calcColumnMeanFromCSV(filename,column):
    """
    Calculates the Mean of a column given as number from a CSV file.

    :param filename: name of csv-file (str)
    :param column:   number of column (int)
    :return:         mean of the column (float)
    """
    with open(filename, "r") as f:
        content = f.readlines()

    del content[0]  # we are not using the column names here, so we delete them

    partLen = len(content[0].split(","))
    if column > partLen: # check if column exists in file
        print("The desired columnnumber %i does not exist in the file which has just %i columns."%(column,partLen))
        sys.exit(1)

    sum = 0
    for line in content:
        useElement = line.split(",")[column-1].rstrip() # just use the desired column and get rid of newline character
        try:
            useElement = float(useElement)
        except:
            print("The Column you would like to know the mean from does not contain numbers.")
            sys.exit(1)
        sum += useElement

    mean = np.divide(sum,len(content))
    return mean


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='File')
    parser.add_argument(type=str, dest="file",
                        help='Provide the filename.')

    parser.add_argument(type=int, dest="column",
                        help='Provide the column of which you want to calculate the mean from. ' +
                             'The left most column has the number 1.')

    file = parser.parse_args().file
    column = parser.parse_args().column

    mean = calcColumnMeanFromCSV(file,column)
    print("The mean of column %i is %f" %(column,mean))

File: Blatt_1/3-csv.R
#############################
# Implemention of function
computeMeanCSV = function(filename, column){

    #Calculates the Mean of a column given as number from a CSV file.
    #
    #   param filename: name of csv-file (str)
    #   param column:   number of column (int)
    #   return:         mean of the column (float)


    # Input
    csv_file = file(filename, open="r")
    csv_lines = readLines(csv_file)
    close(csv_file)

    # Data processing
    csv_lines = csv_lines[-1] # remove header of file (first line)

    csv_lines = strsplit(csv_lines, ",") # split lines seperated by ","

    # sum the requested column
    sum = 0
    for(line in csv_lines){
        sum = sum + as.numeric(line[column])
    }

    mean_column = sum/length(csv_lines)

    # Output
    return(mean_column)
}

#############################
# main part of script
args = commandArgs(trailingOnly = TRUE)

filename = args[1]
column = as.numeric(args[2])

if((length(args)==2) & (!is.na(column))){
    result = computeMeanCSV(filename, column)

    print(paste(c("The mean of the column", column, 'is', result), collapse=" "))
} else{
    print("ERROR: The first argument has to be the filename and the second the integer column number.")
}
